### PR TITLE
Stop copying LogicalPlan and Exprs in `DecorrelatePredicateSubquery`

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Part of https://github.com/apache/datafusion/issues/9637
Closes https://github.com/apache/datafusion/issues/10289 

## Rationale for this change
Reduce copying in 

## What changes are included in this PR?

1. Avoid clones in `DecorrelatePredicateSubquery` 

## Are these changes tested?

covered by existing tests

## Are there any user-facing changes?
No

### CODE REVIEW

File: datafusion/optimizer/src/decorrelate_predicate_subquery.rs [hunk 1]
```diff
@@ -103,62 +106,85 @@ impl DecorrelatePredicateSubquery {
...
+        let LogicalPlan::Filter(filter) = plan else {
+            return Ok(Transformed::no(plan));
+        };
+
+        // if there are no subqueries in the predicate, return the original plan
+        let has_subqueries = split_conjunction(&filter.predicate)
+            .iter()
+            .any(|expr| matches!(expr, Expr::InSubquery(_) | Expr::Exists(_)));
+        if !has_subqueries {
+            return Ok(Transformed::no(LogicalPlan::Filter(filter)));
```
Comment (alamb) [inline datafusion/optimizer/src/decorrelate_predicate_subquery.rs:133]: in the common case where there are no subqueries there will be no rewriting / cloning (which is the same as main

File: datafusion/optimizer/src/decorrelate_predicate_subquery.rs [hunk 2]
```diff
@@ -103,62 +106,85 @@ impl DecorrelatePredicateSubquery {
...
-                    self.extract_subquery_exprs(&filter.predicate, config)?;
-                if subqueries.is_empty() {
-                    // regular filter, no subquery exists clause here
-                    return Ok(None);
-                }
+        internal_err!("Should have called DecorrelatePredicateSubquery::rewrite")
+    }
 
-                // iterate through all exists clauses in predicate, turning each into a join
-                let mut cur_input = filter.input.as_ref().clone();
```
Comment (alamb) [inline datafusion/optimizer/src/decorrelate_predicate_subquery.rs:119]: this is one avoided clone of the input plan

File: datafusion/optimizer/src/decorrelate_predicate_subquery.rs [hunk 3]
```diff
@@ -103,62 +106,85 @@ impl DecorrelatePredicateSubquery {
...
-                    {
-                        cur_input = plan;
-                    } else {
-                        // If the subquery can not be converted to a Join, reconstruct the subquery expression and add it to the Filter
-                        let sub_query_expr = match subquery {
-                            SubqueryInfo {
-                                query,
-                                where_in_expr: Some(expr),
-                                negated: false,
-                            } => in_subquery(expr, query.subquery.clone()),
```
Comment (alamb) [inline datafusion/optimizer/src/decorrelate_predicate_subquery.rs:132]: these are clones of `Arc`s so while this PR removes them and it is less cloning realistically it doesn't matter

File: datafusion/physical-expr/src/scalar_function.rs [hunk 4]
```diff
@@ -37,7 +37,7 @@ use std::sync::Arc;
 use arrow::datatypes::{DataType, Schema};
 use arrow::record_batch::RecordBatch;
 
-use datafusion_common::Result;
+use datafusion_common::{internal_err, Result};
```
Comment (alamb) [inline datafusion/physical-expr/src/scalar_function.rs:40]: This is the change from https://github.com/apache/datafusion/pull/10317 to get the CI to pass

**Review by alamb — COMMENTED** [review]
I think it is easier to see what is actually changed in this PR using whitespace blind diff: https://github.com/apache/datafusion/pull/10318/files?w=1

**Review by waynexia — APPROVED** [review]
LGTM, thanks! Except for avoiding clones, using `if let` to reduce one nest block is also great 👍

**Comment by alamb** [comment]
Thanks for the review @waynexia

**Pull request merged**